iothub_client/samples/pnp/pnp_temperature_controller/pnp_temperature_controller.c

```diff
@@ -0,0 +1,381 @@
+static bool g_hubClientTraceEnabled = true;
+
+// DTMI indicating this device's ModelId.
+static const char g_thermostatModelId[] = "dtmi:com:example:TemperatureController;1";
+
+// PNP_THERMOSTAT_COMPONENT_HANDLE represent the thermostat components that are sub-components of the temperature controller.
+// Note that we do NOT have an analogous DeviceInfo component handle because there is only DeviceInfo subcomponent and its
+// implementation is straightforward.
+PNP_THERMOSTAT_COMPONENT_HANDLE g_thermostatHandle1;
+PNP_THERMOSTAT_COMPONENT_HANDLE g_thermostatHandle2;
+
+// Name of subcomponents that TemmperatureController implements.
+static const char g_thermostatComponent1Name[] = "thermostat1";
+static const size_t g_thermostatComponent1Size = sizeof(g_thermostatComponent1Name) - 1;
+static const char g_thermostatComponent2Name[] = "thermostat2";
+static const size_t g_thermostatComponent2Size = sizeof(g_thermostatComponent2Name) - 1;
+static const char g_deviceInfoComponentName[] = "deviceInformation";
+
+static const char* g_modeledComponents[] = {g_thermostatComponent1Name, g_thermostatComponent2Name, g_deviceInfoComponentName};
+static size_t g_numModeledComponents = sizeof(g_modeledComponents) / sizeof(g_modeledComponents[0]);
+
+// Command implemented by the TemperatureControl component itself to implement reboot.
+static const char g_rebootCommand[] = "reboot";
+
+// An empty JSON body for PnP command responses
+static const char g_emptyJson[] = "{}";
+static const size_t g_emptyJsonSize = sizeof(g_emptyJson) - 1;
+
+// Minimum value we will return for working set, + some random number
+static const int g_workingSetMinimum = 1000;
+// Random number for working set will range between the g_workingSetMinimum and (g_workingSetMinimum+g_workingSetRandomModulo)
+static const int g_workingSetRandomModulo = 500;
+// Format string for sending a telemetry message with the working set.
+static const char g_workingSetTelemetryFormat[] = "{\"workingSet\":%d}";
+
+// Name of the serial number property as defined in this component's DTML
+static const char g_serialNumberPropertyName[] = "serialNumber";
+// Value of the serial number.  NOTE: This must be a legal JSON string which requires value to be in "..."
+static const char g_serialNumberPropertyValue[] = "\"serial-no-123-abc\"";
+
+//
+// PnP_TempControlComponent_InvokeRebootCommand processes the reboot command on the root interface
+//
+static int PnP_TempControlComponent_InvokeRebootCommand(JSON_Value* rootValue, unsigned char** response, size_t* responseSize)
+{
+    int result;
+
+    if (json_value_get_type(rootValue) != JSONNumber)
+    {
+        LogError("Delay payload is not a number");
+        result = PNP_STATUS_BAD_FORMAT;
+    }
+    else
+    {
+        // See caveats section in ../readme.md; we don't actually respect the delay value to keep the sample simple.
+        int delayInSeconds = (int)json_value_get_number(rootValue);
```

Comment:
If someone put in a delay of 1 hour, it means the thread is going to hang there for an hour.  C has a single threaded dispatcher so it would block any future callbacks through that time as well, which is why we tell folks not to block on callbacks.  I'd rather just have a known & doc'd caveat than demonstrate something we tell folks not to do.